solr/core/src/java/org/apache/solr/core/FileSystemConfigSetService.java

```diff
@@ -238,9 +238,14 @@ public List<String> getAllConfigFiles(String configName) throws IOException {
           @Override
           public FileVisitResult visitFile(Path file, BasicFileAttributes attrs)
               throws IOException {
-            // don't include hidden (.) files
-            if (!Files.isHidden(file)) {
-              filePaths.add(configDir.relativize(file).toString());
+            Path filePath = configDir.relativize(file);
+            String filePathStr = filePath.toString();
+            filePathStr =
+                filePathStr.replace(
+                    filePath.getFileSystem().getSeparator(), "/"); // normalize slashes
+            // don't include .metadata.json hidden file
+            if (!filePathStr.equals(METADATA_FILE)) {
+              filePaths.add(filePathStr);
```

Comment:
There might be other subdirs  which would have a separator, not only METADATA_FILE.

Reply:
Not sure I understand. In `visitFile` you end up with a file path. Why do we need to do anything about the slashes if we stick with Path? 

Why are we storing strings for file paths? We should be storing `Path` only and comparing paths. This avoids the whole string manipulation of file paths.

Reply:
I hear you on embracing Path.  In this particular case, the ConfigSetService API (that is being implemented here) is to return a List<String> not a List<Path>.  The ConfigSetService cannot return Paths because the data isn't necessarily on the file system.  And I've never seen a Path/FileSystem implementation for ZK; it's hypothetical.

Reply:
Hmmm I see what you mean about `List<String>` I guess I'm confused maybe by the variable name and what is going on here. I haven't looked at the original PR or the whole change just related to this error.

I'll try to lay out my thoughts and maybe there is some misunderstanding I have.

* normalizing paths to `/` - if we do this - then on Windows this is no longer a filesystem path right? We are forcing these to be what? just strings to compare?
* In my head, we don't need to normalize paths - since `FileSystemConfigSetService` is run per node. Even if you had a mixed Windows/Linux Solr setup - each node (either Windows or Linux) would only care about file paths in its format (`/` or `\`). The strings being compared should just be comparable regardless of this "normalizing" - if its a path on windows - who cares if its `\` separating. 
* We are looking at filePathStr for `METADATA_FILE` - which should be just a `Path` comparison - regardless of what ends up in `filePaths` right?

Reply:
@NazerkeBS / @dsmiley any update here? Windows tests are still failing.

Reply:
I'm putting together a PR to show what I mean about using Path instead of string manipulation. I'll have it up shortly.

Reply:
Opened https://github.com/apache/solr/pull/1278

Reply:
Well https://github.com/apache/solr/pull/1278 fails since the test is testing both ZK and Filesystem. ZK always returns `/` and Filesystem doesn't. I'm not sure the actual string values matter but the test is trying to assert that ZK and Filesystem should return the same values. Not sure which is correct or not.

Reply:
This ought to be documented in the method contract for getAllConfigFiles() better, although at least the existing docs give examples using `/` as a separator.  I see two users of this method:
* `org.apache.solr.core.backup.BackupManager#downloadConfigToRepo` which takes these and calls `org.apache.solr.core.backup.repository.BackupRepository#resolve` which isn't documented clearly on separators either.  Its file system based implementation assumes it's slash-compatible, which means it'll break if getAllConfigFiles() were to normalize them.
* `org.apache.solr.handler.configsets.UploadConfigSetAPI#uploadConfigSet` which will do two things:
  * compare the result with a zip entry file name.  I dug a little and[ Zip files should normalize to '/' separator](https://stackoverflow.com/a/44387973).
  * it'll call `org.apache.solr.core.ConfigSetService#deleteFilesFromConfig` which doesn't document this separator matter either.  

So it's a bit of a mess.  I think getAllConfigFiles() should normalize to '/' due to the generic nature of the abstraction (isn't specific to a local file system) and ensure that `org.apache.solr.core.backup.repository.BackupRepository#resolve` and  `org.apache.solr.core.ConfigSetService#deleteFilesFromConfig` translate a `/` in its input a file system native separator if there's a difference.

Reply:
I think I addressed these in 086ff2cf02e88757c18cf517491f4a740706e3c2 - added comments to the javadocs and cleaned up the places in question to either normalize or not to either '/' or the OS file path separator.